File: services/worker-image/internal/broker/broker.go
// Package broker is the worker's RabbitMQ adapter: it declares the (shared,
// idempotent) topology, consumes jobs with bounded retries via a dead-letter +
// TTL retry loop, and publishes lifecycle events.
package broker

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	amqp "github.com/rabbitmq/amqp091-go"
)

const (
	ExchangeJobs    = "media.jobs"
	ExchangeEvents  = "media.events"
	ExchangeRetry   = "media.jobs.dlx"
	ExchangeParking = "media.jobs.parking"

	QueueImage   = "q.image"
	QueueOCR     = "q.ocr"
	QueueRetry   = "q.retry"
	QueueParking = "q.parking"
)

// Job is the wire contract published by the gateway on media.jobs. It mirrors
// api-gateway/internal/media.Job — keep the JSON tags in sync.
type Job struct {
	ID         string   `json:"job_id"`
	Kind       string   `json:"kind"`
	Status     string   `json:"status"`
	SourceKey  string   `json:"source_key"`
	SourceMIME string   `json:"source_mime"`
	SizeBytes  int64    `json:"size_bytes"`
	Operations []string `json:"operations"`
}

// Event is emitted on media.events as a job changes state.
type Event struct {
	JobID     string `json:"job_id"`
	Kind      string `json:"kind"`
	Status    string `json:"status"`
	Message   string `json:"message,omitempty"`
	Progress  int    `json:"progress"`
	Artifact  string `json:"artifact,omitempty"`
	Timestamp int64  `json:"ts"`
}

type Broker struct {
	conn       *amqp.Connection
	ch         *amqp.Channel
	maxRetries int
}

func Connect(url string, prefetch, maxRetries int, retryTTL time.Duration) (*Broker, error) {
	conn, err := amqp.Dial(url)
	if err != nil {
		return nil, fmt.Errorf("dial rabbitmq: %w", err)
	}
	ch, err := conn.Channel()
	if err != nil {
		_ = conn.Close()
		return nil, fmt.Errorf("open channel: %w", err)
	}
	if err := ch.Qos(prefetch, 0, false); err != nil {
		_ = conn.Close()
		return nil, fmt.Errorf("set qos: %w", err)
	}
	b := &Broker{conn: conn, ch: ch, maxRetries: maxRetries}
	if err := b.declareTopology(retryTTL); err != nil {
		_ = conn.Close()
		return nil, err
	}
	return b, nil
}

func (b *Broker) declareTopology(retryTTL time.Duration) error {
	for _, ex := range []string{ExchangeJobs, ExchangeEvents, ExchangeRetry, ExchangeParking} {
		if err := b.ch.ExchangeDeclare(ex, amqp.ExchangeTopic, true, false, false, false, nil); err != nil {
			return fmt.Errorf("declare exchange %s: %w", ex, err)
		}
	}
	workArgs := amqp.Table{"x-queue-type": "quorum", "x-dead-letter-exchange": ExchangeRetry}
	for queue, key := range map[string]string{QueueImage: "image.process", QueueOCR: "ocr.extract"} {
		if _, err := b.ch.QueueDeclare(queue, true, false, false, false, workArgs); err != nil {
			return fmt.Errorf("declare queue %s: %w", queue, err)
		}
		if err := b.ch.QueueBind(queue, key, ExchangeJobs, false, nil); err != nil {
			return fmt.Errorf("bind queue %s: %w", queue, err)
		}
	}
	retryArgs := amqp.Table{"x-dead-letter-exchange": ExchangeJobs, "x-message-ttl": int32(retryTTL.Milliseconds())}
	if _, err := b.ch.QueueDeclare(QueueRetry, true, false, false, false, retryArgs); err != nil {
		return fmt.Errorf("declare retry queue: %w", err)
	}
	if err := b.ch.QueueBind(QueueRetry, "#", ExchangeRetry, false, nil); err != nil {
		return fmt.Errorf("bind retry queue: %w", err)
	}
	if _, err := b.ch.QueueDeclare(QueueParking, true, false, false, false, nil); err != nil {
		return fmt.Errorf("declare parking queue: %w", err)
	}
	if err := b.ch.QueueBind(QueueParking, "#", ExchangeParking, false, nil); err != nil {
		return fmt.Errorf("bind parking queue: %w", err)
	}
	return nil
}

// Handler processes a single job. Returning an error triggers the retry/DLQ path.
type Handler func(ctx context.Context, job Job, attempt int) error

// Consume runs the delivery loop on the given queue until ctx is cancelled.
// Retry policy:
//   - success            -> ack
//   - failure, attempt<N  -> nack(requeue=false) => dead-letters to retry queue,
//     which TTLs the message back onto media.jobs for another attempt
//   - failure, attempt>=N -> republish to the parking exchange, then ack
func (b *Broker) Consume(ctx context.Context, queue string, h Handler) error {
	deliveries, err := b.ch.Consume(queue, "", false, false, false, false, nil)
	if err != nil {
		return fmt.Errorf("consume %s: %w", queue, err)
	}
	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case d, ok := <-deliveries:
			if !ok {
				return fmt.Errorf("delivery channel closed")
			}
			b.handle(ctx, d, h)
		}
	}
}

func (b *Broker) handle(ctx context.Context, d amqp.Delivery, h Handler) {
	attempt := deathCount(d.Headers) + 1

	var job Job
	if err := json.Unmarshal(d.Body, &job); err != nil {
		// Unparseable: send straight to the parking lot, it will never succeed.
		_ = b.park(ctx, d, "unmarshal: "+err.Error())
		_ = d.Ack(false)
		return
	}

	if err := h(ctx, job, attempt); err != nil {
		if attempt >= b.maxRetries {
			_ = b.park(ctx, d, err.Error())
			_ = d.Ack(false) // terminal — remove from work queue
			return
		}
		_ = d.Nack(false, false) // dead-letter into the retry loop
		return
	}
	_ = d.Ack(false)
}

func (b *Broker) park(ctx context.Context, d amqp.Delivery, reason string) error {
	headers := amqp.Table{"x-parking-reason": reason}
	return b.ch.PublishWithContext(ctx, ExchangeParking, d.RoutingKey, false, false, amqp.Publishing{
		ContentType:  "application/json",
		DeliveryMode: amqp.Persistent,
		Headers:      headers,
		Body:         d.Body,
	})
}

// PublishEvent emits a lifecycle event on media.events keyed by job kind.
func (b *Broker) PublishEvent(ctx context.Context, e Event) error {
	e.Timestamp = time.Now().UnixMilli()
	body, _ := json.Marshal(e)
	return b.ch.PublishWithContext(ctx, ExchangeEvents, "event."+e.Kind, false, false, amqp.Publishing{
		ContentType:  "application/json",
		DeliveryMode: amqp.Transient,
		Body:         body,
	})
}

func (b *Broker) Close() error {
	if b.conn != nil {
		return b.conn.Close()
	}
	return nil
}

// deathCount reads the number of prior dead-letterings from the x-death header
// RabbitMQ maintains, which is how we count retry attempts.
func deathCount(h amqp.Table) int {
	raw, ok := h["x-death"]
	if !ok {
		return 0
	}
	deaths, ok := raw.([]any)
	if !ok || len(deaths) == 0 {
		return 0
	}
	first, ok := deaths[0].(amqp.Table)
	if !ok {
		return 0
	}
	if c, ok := first["count"].(int64); ok {
		return int(c)
	}
	return 0
}

File: services/worker-ocr/mediaforge_ocr/broker.py
"""RabbitMQ adapter for the OCR worker.

Declares the same (idempotent) topology as the Go services, consumes ``q.ocr``
with bounded retries through the dead-letter + TTL retry loop, and publishes
lifecycle events on ``media.events``.
"""

from __future__ import annotations

import json
import logging
from collections.abc import Callable
from dataclasses import dataclass, field

import pika
from opentelemetry import trace
from opentelemetry.propagate import extract, inject
from opentelemetry.trace import SpanKind
from pika.adapters.blocking_connection import BlockingChannel
from pika.spec import Basic, BasicProperties

log = logging.getLogger("mediaforge.ocr.broker")
tracer = trace.get_tracer("mediaforge.ocr.broker")

EXCHANGE_JOBS = "media.jobs"
EXCHANGE_EVENTS = "media.events"
EXCHANGE_RETRY = "media.jobs.dlx"
EXCHANGE_PARKING = "media.jobs.parking"

QUEUE_IMAGE = "q.image"
QUEUE_OCR = "q.ocr"
QUEUE_RETRY = "q.retry"
QUEUE_PARKING = "q.parking"


@dataclass
class Job:
    """Wire contract mirroring the Go gateway's media.Job."""

    job_id: str
    kind: str
    source_key: str
    source_mime: str = ""
    size_bytes: int = 0
    operations: list[str] = field(default_factory=list)

    @staticmethod
    def from_bytes(body: bytes) -> Job:
        d = json.loads(body)
        return Job(
            job_id=d["job_id"],
            kind=d.get("kind", "ocr"),
            source_key=d["source_key"],
            source_mime=d.get("source_mime", ""),
            size_bytes=d.get("size_bytes", 0),
            operations=d.get("operations") or [],
        )


class PermanentError(Exception):
    """A failure no retry can fix (e.g. an unreadable payload).

    Mirrors the Go broker's PermanentError: the consumer parks the message on
    the first attempt instead of cycling it through the retry loop.
    """


# Handler raises to signal failure -> retry/park (PermanentError parks at once).
Handler = Callable[[Job, int], None]


class Broker:
    def __init__(self, url: str, prefetch: int, max_retries: int, retry_ttl_ms: int):
        self._params = pika.URLParameters(url)
        self._prefetch = prefetch
        self._max_retries = max_retries
        self._retry_ttl_ms = retry_ttl_ms
        self._conn = pika.BlockingConnection(self._params)
        self._ch = self._conn.channel()
        self._declare_topology()
        self._ch.basic_qos(prefetch_count=prefetch)

    def _declare_topology(self) -> None:
        for ex in (EXCHANGE_JOBS, EXCHANGE_EVENTS, EXCHANGE_RETRY, EXCHANGE_PARKING):
            self._ch.exchange_declare(ex, exchange_type="topic", durable=True)

        work_args = {"x-queue-type": "quorum", "x-dead-letter-exchange": EXCHANGE_RETRY}
        for queue, key in ((QUEUE_IMAGE, "image.process"), (QUEUE_OCR, "ocr.extract")):
            self._ch.queue_declare(queue, durable=True, arguments=work_args)
            self._ch.queue_bind(queue, EXCHANGE_JOBS, routing_key=key)

        retry_args = {"x-dead-letter-exchange": EXCHANGE_JOBS, "x-message-ttl": self._retry_ttl_ms}
        self._ch.queue_declare(QUEUE_RETRY, durable=True, arguments=retry_args)
        self._ch.queue_bind(QUEUE_RETRY, EXCHANGE_RETRY, routing_key="#")

        self._ch.queue_declare(QUEUE_PARKING, durable=True)
        self._ch.queue_bind(QUEUE_PARKING, EXCHANGE_PARKING, routing_key="#")

    def consume(self, queue: str, handler: Handler) -> None:
        """Block consuming ``queue`` until the connection is closed."""

        def _on_message(
            ch: BlockingChannel,
            method: Basic.Deliver,
            props: BasicProperties,
            body: bytes,
        ) -> None:
            attempt = _death_count(props.headers) + 1

            # Continue the distributed trace: extract the W3C context the gateway
            # injected into the message headers and open a consumer span.
            ctx = extract(props.headers or {})
            with tracer.start_as_current_span(
                f"consume {method.routing_key}",
                context=ctx,
                kind=SpanKind.CONSUMER,
                attributes={
                    "messaging.system": "rabbitmq",
                    "messaging.rabbitmq.destination.routing_key": method.routing_key or "",
                    "messaging.rabbitmq.delivery.attempt": attempt,
                },
            ) as span:
                try:
                    job = Job.from_bytes(body)
                except (ValueError, KeyError) as exc:
                    span.record_exception(exc)
                    self._park(method.routing_key, body, f"unmarshal: {exc}")
                    ch.basic_ack(method.delivery_tag)
                    return

                span.set_attribute("mediaforge.job.id", job.job_id)
                try:
                    handler(job, attempt)
                    ch.basic_ack(method.delivery_tag)
                except Exception as exc:  # noqa: BLE001 — broker decides retry vs park
                    span.record_exception(exc)
                    if isinstance(exc, PermanentError) or attempt >= self._max_retries:
                        self._park(method.routing_key, body, str(exc))
                        ch.basic_ack(method.delivery_tag)
                    else:
                        ch.basic_nack(method.delivery_tag, requeue=False)

        self._ch.basic_consume(queue, _on_message, auto_ack=False)
        self._ch.start_consuming()

    def _park(self, routing_key: str, body: bytes, reason: str) -> None:
        self._ch.basic_publish(
            EXCHANGE_PARKING,
            routing_key,
            body,
            properties=BasicProperties(
                content_type="application/json",
                delivery_mode=2,
                headers={"x-parking-reason": reason},
            ),
        )
        log.warning("parked message: %s", reason)

    def publish_event(self, event: dict) -> None:
        # Inject the current trace context so the realtime-gateway's fan-out
        # joins the same trace.
        headers: dict = {}
        inject(headers)
        self._ch.basic_publish(
            EXCHANGE_EVENTS,
            f"event.{event.get('kind', 'ocr')}",
            json.dumps(event),
            properties=BasicProperties(
                content_type="application/json", delivery_mode=1, headers=headers
            ),
        )

    def close(self) -> None:
        try:
            self._ch.stop_consuming()
        finally:
            self._conn.close()


def _death_count(headers: dict | None) -> int:
    if not headers:
        return 0
    deaths = headers.get("x-death")
    if not deaths:
        return 0
    first = deaths[0]
    return int(first.get("count", 0)) if isinstance(first, dict) else 0
